mlir/lib/Dialect/EmitC/IR/EmitC.cpp

```diff
@@ -227,6 +228,17 @@ LogicalResult emitc::AssignOp::verify() {
 bool CastOp::areCastCompatible(TypeRange inputs, TypeRange outputs) {
   Type input = inputs.front(), output = outputs.front();
 
+  // Arrays & pointers don't cast to/from scalars
+  if (isa<emitc::ArrayType, emitc::PointerType>(input) !=
+      isa<emitc::ArrayType, emitc::PointerType>(output))
+    return false;
+
+  // Arrays & pointers can be casted to each other.
+  if (isa<emitc::ArrayType, emitc::PointerType>(input) &&
+      isa<emitc::ArrayType, emitc::PointerType>(output))
+    return true;
```

Comment:
Cast array to array but maybe not to pointer since we don't do type matching

Reply:
I'd exclude pointer to array and vice versa for now; I don't think that his properly works when casting as reference